### PR TITLE
Feat/streaming

### PR DESCRIPTION
**Pull request by jxsl13**
reference apis from stdlib: https://pkg.go.dev/compress
in order for users to work with this library the same way as one does with the stdlib.
I also added a small benchmark for the comparison of Compress and Writer.Write. 
Compress is still faster in raw operations per time unit, while the Writer has small memory usage and no heap allocations at all.
It's kind of a tradeoff what someone might want to use imo.
For streaming to a connection or a file, it might be easier to do an `io.Copy(huffman.NewWriter(conn),  bytes.NewBuffer(...))`

### CODE REVIEW

File: huffman.go [hunk 1]
```diff
@@ -203,21 +81,18 @@ func (huff *Huffman) Decompress(data []byte) ([]byte, error) {
...
 
 func (huff *Huffman) Compress(data []byte) ([]byte, error) {
-	if huff.initalized == false {
-		huff.Init()
-	}
 
 	srcIndex := 0
 	end := len(data)
 	bits := uint32(0)
-	bitcount := uint32(0)
```
Comment (ChillerDragon) [inline huffman.go:220]: Ou nice does this fix a bug?

File: huffman.go [hunk 2]
```diff
@@ -2,165 +2,43 @@ package huffman
 
 import (
 	"errors"
-	"sort"
 )
 
 const (
-	eofSymbol  = 256
```
Comment (ChillerDragon) [inline huffman.go:9]: why did you rename those? I am not sure I like the Huffman prefix in a huffman package. Seems redundant.

File: testdata/fuzz/FuzzWrite/caf81e9797b19c76 [hunk 3]
```diff
@@ -0,0 +1,2 @@
+go test fuzz v1
```
Comment (ChillerDragon) [inline testdata/fuzz/FuzzWrite/caf81e9797b19c76:1]: what does this file do?

**Review by ChillerDragon — CHANGES_REQUESTED** [review]
Could you please also update the README.md the old example no longer works because now `NewHuffman` is needed. Maybe you can also add an example for the newly added feature.

I am a bit skeptical about the entire pull request. What is its purpose? Are you planning to use streaming in https://github.com/teeworlds-go/protocol? I don't see where this could be used.

Also the code diff is huge. Feels like you just dumped in the code from [twapi](https://github.com/jxsl13/twapi/blob/e06f5677d75afe79df0a91ae9af49b90ed08447e/compression/huffman.go). If that is the case then I do not understand the purpose of keeping the same code in two repos. I would have used your huffman code straight away if i knew it existed when starting the project. I rewrote it from scratch because I thought that was the only way.

File: huffman.go (hunk 1 above)
Comment (jxsl13) [reply huffman.go:220]: no it's one just doesn't need 32 bits for that type.

File: huffman.go (hunk 2 above)
Comment (jxsl13) [reply huffman.go:9]: will remove the prefix

File: testdata/fuzz/FuzzWrite/caf81e9797b19c76 (hunk 3 above)
Comment (jxsl13) [reply testdata/fuzz/FuzzWrite/caf81e9797b19c76:1]: it is a file that is generated from the fuzz tests. It is used for regression testing as input data for the FizzWrite test.

**Comment by jxsl13** [comment]
on the long run, I'd like to discontinue twapi and participate here, which is why, yes I do take some stuff from twapi, but that's not completely the case here. There is no streaming in twapi, at least not one that conforms to the standard library interfaces like io.Writer.

The streaming could potentially be used for the protocol library but also for anyone wanting to use it for their stuff which might be completely out of scope of this teeworlds-go organization.

Because [net.Conn](https://pkg.go.dev/net#Conn) also implements the io.Reader/Writer interface, this Writer implementation would for example make it possible to directly stream the compressed data to the connection using 
`io.Copy(huffman.NewWriter(conn),  bytes.NewBuffer(someByteSliceBufferFromSnap))`

**Comment by ChillerDragon** [comment]
> on the long run, I'd like to discontinue twapi and participate here


Maybe I should warn you that I am not planning to actively maintain any repository in this organization. I'm just on a few weeks streak to see how far I can get. And hope some day someone will find it useful. After that this project will probably end up on my pile of inactive projects. Possibly receiving basic life support.

File: writer.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+			err = h.append(byte(bits))
+			if err != nil {
+				return
+			}
+			bits >>= 8
+			bitCount -= 8
+		}
+	}
+
+	nodeEOF := h.d.nodes[EOFSymbol]
```
Comment (ChillerDragon) [inline writer.go]: I would like to keep the old strict camel casing for EOF here. Since ddnet went through heinrichs style refactor I started to appreciate strict camel casing no matter what. Eases the brain by taking away one more thought.

Comment (jxsl13) [reply writer.go]: I personally do not care about that too much but Go is also pretty opinionated about correct camel casing which is why I'd like to share this small section from the official docs: https://go.dev/wiki/CodeReviewComments#initialisms

if you still wanna go back to the the previous one, I will change that, no issue at all.

Comment (ChillerDragon) [reply writer.go]: I like following the official ideals. But as said on discord I think its worth here to violate it. Just to be consistent with ddnet. I just cba to type ClientId in C++ code and then ClientID in Go.

**Review by ChillerDragon — APPROVED** [review]
